### DISCUSSION TITLE
PICO lightsleep and pin hold

### DISCUSSION THREAD

**Discussion by davefes** [micropython/micropython#12648] [RP2040 / Pico] [answered]
One benefit of lightsleep, on the ESP32, is that one can hold the pin state during lightsleep.  This capability does not appear to be possible on the RP2040 devices.  Have I missed something?

**Comment by GitHubsSilverBullet** [comment, marked as answer]
What makes you think it doesn't?
```python


from machine import Pin, lightsleep

l = [x for x in range(50,200+1,10)]
idx = 0

pin = Pin(25,Pin.OUT)
while True:
    pin.toggle()
    lightsleep(l[idx])
    pin.toggle()
    lightsleep(l[idx])
    idx = (idx+1) % len(l)
```

**Comment by davefes** [comment]
I have since found a comment `RP2040 dormant mode has all the RAM still powered`, thanks Google.

And thank you for the code snippet.

**Comment by davefes** [comment]
I tried the code snippet and now I can't reclaim my PICO.  Holding the boot selector while plugging it in says the .uf2 file is no longer there, but the program still runs and I can't stop it.

**Reply by GitHubsSilverBullet** [reply]
Why on earth would you put something with an infinite loop into boot.py?
At least an emergency pin input should have been defined.
Regarding the uf2 file, it'll never show up after plugging in. It's written to FLASH and that's it.
Just reflash and try again.

**Reply by robert-hh** [reply]
the .uf2 file disappears after firmware update is finished. If you have placed the code snipped in main.py or boot.py, it is indeed hard to stop. USB deregisters during lightsleep. What you could do is wipe out the flash. There is a small .uf2 file for this purpose. https://datasheets.raspberrypi.com/soft/flash_nuke.uf2
Load that, and it will clear the flash, firmware and files. Then you can reload the MicroPython firmware again.

**Reply by davefes** [reply]
> Why on earth would you put something with an infinite loop into boot.py?

I put it in main.py and because I didn't stop and think!

Thank you @robert-hh for a get out of jail solution.

**Reply by robert-hh** [reply]
You're welcome. Once I have been in that jail too. The Pycom MicroPython port has the feature, that if a certain Pin is high at boot time, boot.py and main.py will not be executed. On some ports, you can implement a similar behavior by using the feature, that main,py will not be executed if boot.py fails. So if you check the level of a pin of your choice at the start of boot.py, and if it has a `do not run` condition, you can then raise an exception. Then, main.py will not run.
Unfortunately that is not implemented at the RP2 port. Can easily be added. I'll make a PR. Until then, you could for instance rename main.py.

**Reply by davefes** [reply]
I normally put a 2-5 second sleep at the beginning to give me a chance of breaking-in.  I was too fast with the copy/paste :(

**Reply by peterhinch** [reply]
I do the same, but it doesn't make sense for micro-power applications that use deepsleep because the delay consumes power each time you wake. The solution I use is similar to that of @robert-hh  but it doesn't require a PR:
```python
from machine import Pin
from sys import exit
breakout = Pin(16, Pin.IN, Pin.PULL_UP)
if not breakout():  # Pull down for REPL
    exit()
```

**Reply by davefes** [reply]
Good point.  I have read through @glenn20 mods for reducing current, but haven't had a challenging enough application ... yet.

Thanks for the code snippet.

**Reply by robert-hh** [reply]
Of course you can put that to the start of main.py. But since this mechanism of skipping main.py if boot.py fails is implemented on a few ports, I'll make a PR to extend it to all major ports.

**Comment by davefes** [comment]
@peterhinch, tried your suggestion as follows:
```py
from time import sleep
from machine import Pin
from sys import exit


breakout = Pin(16, Pin.IN, Pin.PULL_UP)

if not breakout():  # Pull down for REPL
    exit()

led = Pin('LED', Pin.OUT)
led.off()

while True:
    led.off()
    sleep(0.5)
    led.on()
    sleep(0.5)
```
It works, but I have noticed an unexpected behaviour.  I will start another thread.

**Reply by GitHubsSilverBullet** [reply]
I'd suggest a tiny delay between the `breakout = Pin(…)` and the `if not breakout()`.
Depending on electrical charges an immediate check might not work as expected due to the high resistance pull-up (50kΩ…80kΩ)

PS. On reset the RP2040 has all inputs pulled down to ›low‹ and any existing capacitance is therefore discharged. Toggling the input to ›pull-up‹ will allow the capacitance to charge, but it might take a couple of microseconds to reach the Schmitt-Trigger threshold and thus (if read to fast) could still be read as ›low‹.

PPS. The inverse is true for capacitance between input and Vcc. They're fully charged and need to discharge, thus amplifying the "problem".

**Reply by davefes** [reply]
Sorry, I started another thread where I discovered that a short delay was required.
